aptos-move/framework/src/natives/cryptography/algebra/arithmetics/neg.rs

```diff
@@ -0,0 +1,77 @@
+// Copyright © Aptos Foundation
+
+use crate::{
+    abort_unless_arithmetics_enabled_for_structure, abort_unless_feature_flag_enabled,
+    ark_unary_op_internal,
+    natives::{
+        cryptography::algebra::{
+            abort_invariant_violated, feature_flag_from_structure, gas::GasParameters,
+            AlgebraContext, Structure, MOVE_ABORT_CODE_NOT_IMPLEMENTED,
+        },
+        helpers::{SafeNativeContext, SafeNativeError, SafeNativeResult},
+    },
+    safe_borrow_element, safely_pop_arg, store_element, structure_from_ty_arg,
+};
+use ark_ff::Field;
+use move_core_types::gas_algebra::NumArgs;
+use move_vm_types::{loaded_data::runtime_types::Type, values::Value};
+use smallvec::{smallvec, SmallVec};
+use std::{collections::VecDeque, ops::Neg, rc::Rc};
+
+pub fn neg_internal(
+    gas_params: &GasParameters,
+    context: &mut SafeNativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    assert_eq!(1, ty_args.len());
+    let structure_opt = structure_from_ty_arg!(context, &ty_args[0]);
+    abort_unless_arithmetics_enabled_for_structure!(context, structure_opt);
+    match structure_opt {
+        Some(Structure::BLS12381Fr) => ark_unary_op_internal!(
+            context,
+            args,
+            ark_bls12_381::Fr,
+            neg,
+            gas_params.ark_bls12_381_fr_neg * NumArgs::one()
+        ),
+        Some(Structure::BLS12381Fq12) => ark_unary_op_internal!(
+            context,
+            args,
+            ark_bls12_381::Fq12,
+            neg,
+            gas_params.ark_bls12_381_fq12_neg * NumArgs::one()
+        ),
+        Some(Structure::BLS12381G1Affine) => ark_unary_op_internal!(
+            context,
+            args,
+            ark_bls12_381::G1Projective,
+            neg,
+            gas_params.ark_bls12_381_g1_proj_neg * NumArgs::one()
+        ),
+        Some(Structure::BLS12381G2Affine) => ark_unary_op_internal!(
+            context,
+            args,
+            ark_bls12_381::G2Projective,
+            neg,
+            gas_params.ark_bls12_381_g2_proj_neg * NumArgs::one()
+        ),
+        Some(Structure::BLS12381Gt) => {
+            let handle = safely_pop_arg!(args, u64) as usize;
+            safe_borrow_element!(context, handle, ark_bls12_381::Fq12, element_ptr, element);
+            context.charge(gas_params.ark_bls12_381_fq12_inv * NumArgs::one())?;
+            let new_element = match element.inverse() {
+                Some(e) => e,
+                None => {
+                    abort_invariant_violated();
```

Comment:
This should never happen unless either of the following happens:
1. There's an error inside `arkworks`
2. We messed up & somehow ended up adding a non-GT element to the algebra context.

Nonetheless, we do not want to add code that panics in a native, no matter how sure we are "it will never happen."

AFAIU from @vgao1996, this could be a situation where we want to return a `SafeNativeError::InvariantViolation(UNKNOWN_INVARIANT_VIOLATION_ERROR)`.

We could also abort, but that would require defining a new, **unused** abort code, because we (believe we) can never even trigger the path that returns this new abort code. So we'd be defining it without ever testing for it.

Reply:
Fixed.